docs/user-manual/core-tools.md

```diff
@@ -219,12 +219,354 @@ As c++ is a strongly typed language, the signal template parameters signature ha
+
+<div class="warning-msg">
+  <i class="fa fa-warning"></i>
+  Don't forget to export the referenced environment variables before using the XML parser as you will get an exception if they are not defined !
+</div>
+
+### XML for loops
+
+The element `<for>` allows to run a simple for loop and duplicate XML elements. To keep it simple, the loop is run over integer id (`id`) from a lower value (`begin`) to an upper value (`end`) with a user defined increment (`increment`). A for loop is run over the whole content of the `<for>` element, look for the string "$FOR{*id*}" and replace it by the current loop value. The replacement operates on:
+
+- the XML comments content
+- all the XML element attributes
+- the XML texts
+
+Example:
+```xml
+<dqm4hep>
+  <for id="count" begin="0" end="3" increment="1">
+    <!-- The current value of the counter is $FOR{count} -->
+    <counter value="$FOR{count}"/>
+  </for>
+</dqm4hep>
+```
+
+will be replaced by:
+```xml
+<dqm4hep>
+  <!-- The current value of the counter is 0 -->
+  <counter value="0"/>
+  <!-- The current value of the counter is 1 -->
+  <counter value="1"/>
+  <!-- The current value of the counter is 2 -->
+  <counter value="2"/>
+  <!-- The current value of the counter is 3 -->
+  <counter value="3"/>
+</dqm4hep>
+```
+
+By construction, it is also possible the run nested for loops by changing the loop id:
+
+```xml
+<dqm4hep>
+  <for id="x" begin="0" end="2" increment="1">
+    <for id="y" begin="0" end="2" increment="1">
+      <entry x="$FOR{x}" y="$FOR{y}"/>
+    </for>
+  </for>
+</dqm4hep>
+```
+
+will be replaced by:
+
+```xml
+<dqm4hep>
+  <entry x="0" y="0"/>
+  <entry x="0" y="1"/>
+  <entry x="0" y="2"/>
+  <entry x="1" y="0"/>
+  <entry x="1" y="1"/>
+  <entry x="1" y="2"/>
+  <entry x="2" y="0"/>
+  <entry x="2" y="1"/>
+  <entry x="2" y="2"/>
+</dqm4hep>
+```
+
+The `<for>` element attributes possible values are summurized in table below. 
+ 
+Attribute | Type    | Optional ? | Default value
+--------- | ------- | ---------- | -------------
+id        | string  | false      | - 
+begin     | integer | false      | - 
+end       | integer | false      | -
+increment | integer | true       | 1
+
+## XML parsing ordering
+
+The XML file is processed in the given order:
+
+- Includes resolution
+- Constants and environment variables
+- Database connections and parameter select
+- For loops
+
+This has the following effects:
+
+1. Constants and environment variables can be used in database connections and for loops
+2. As includes are processed before constants, the `ref` attribute of the `<include>` element can not refer to any constant. In this case the constant will not replaced and the file will not been included.
+3. As constants are processed after includes, it possible to define constants in an included file. The `<constants>` section in the included file has to be defined in the root element and the `<include>` element in the main file has to be defined in the root element.
```

Comment:
As constants are processed after includes, it **is** possible to define constants in an included file.